controllers/containerimagebuild_controller.go

```diff
@@ -99,3 +103,51 @@ func (r *ContainerImageBuildReconciler) Reconcile(req ctrl.Request) (ctrl.Result
 
 	return ctrl.Result{}, nil
 }
+
+func (r *ContainerImageBuildReconciler) RunGC(retentionCount int) {
+	ctx := context.Background()
+	log := r.Log.WithName("GC")
+	log.Info("Launching cleanup operation")
+
+	list := &forgev1alpha1.ContainerImageBuildList{}
+	if err := r.List(ctx, list); err != nil {
+		log.Error(err, "Failed to list build resources, something may be wrong")
+		return
+	}
+
+	listLen := len(list.Items)
+	if listLen == 0 {
+		log.V(1).Info("No build resources found, aborting")
+		return
+	}
+	log.Info("Fetched all build resources", "count", listLen)
+
+	log.V(1).Info("Filtering builds by state", "states", []forgev1alpha1.BuildState{
+		forgev1alpha1.BuildStateCompleted, forgev1alpha1.BuildStateFailed,
+	})
+	var builds []forgev1alpha1.ContainerImageBuild
+	for _, cib := range list.Items {
+		state := cib.Status.State
+		if state == forgev1alpha1.BuildStateCompleted || state == forgev1alpha1.BuildStateFailed {
+			builds = append(builds, cib)
+		}
+	}
+
+	if len(builds) <= retentionCount {
+		log.Info("Total resources are less than or equal to retention limit, aborting", "resourceCount", len(builds), "retentionCount", retentionCount)
+		return
+	}
+	log.Info("Total resources eligible for deletion", "count", len(builds))
+
+	sort.Slice(builds, func(i, j int) bool {
+		return builds[i].CreationTimestamp.Before(&builds[j].CreationTimestamp)
+	})
+
+	for _, build := range builds[:len(builds)-retentionCount] {
+		if err := r.Delete(ctx, &build); err != nil {
```

Comment:
Background or foreground? I had this debate with myself inside domino, but I'm curious your thoughts. I think the default is to block and there's no real performance concerns since this loop runs infrequently.

Reply:
Foreground for now -> https://github.com/dominodatalab/forge/pull/49/commits/c3933061c019b0b83c3b42bc7e3c1767043ff1d6